Update README to mention the usage of feature-based injection (variant service) (#388)
microsoft/FeatureManagement-Dotnet: README.md

* [Targeting](#targeting)
  * [Targeting Exclusion](#targeting-exclusion)
* [Variants](#variants)

> * [Variants in Dependency Injection](#variants-in-dependency-injection)

Reviewer: Sub bullet.

| `DefaultWhenDisabled` | Specifies which variant should be used when a variant is requested while the feature is considered disabled. |
| `DefaultWhenEnabled` | Specifies which variant should be used when a variant is requested while the feature is considered enabled and no other variant was assigned to the user. |
| `User` | Specifies a variant and a list of users to whom that variant should be assigned. | 

> | `Group` | Specifies a variant and a list of groups the current user has to be in at least one of for that variant to be used. |

Reviewer 2: ```suggestion
| `Group` | Specifies a variant and a list of groups. The variant will be assigned if the user is at least one of the groups.|
```

Allocation logic is similar to the [Microsoft.Targeting](./README.md#MicrosoftTargeting) feature filter, but there are some parameters that are present in targeting that aren't in allocation, and vice versa. The outcomes of targeting and allocation are not related.

> **Note:** To allow allocating feature variants, you need to register `ITargetingContextAccessor`. This could be done by calling `WithTargeting<T>` method.

Reviewer: ```suggestion
**Note:** To allow allocating feature variants, you need to register `ITargetingContextAccessor`. This can be done by calling the `WithTargeting<T>` method.

```

### Overriding Enabled State with a Variant

> You can use variants to override the enabled state of a feature flag. This gives variants an opportunity to extend the evaluation of a feature flag. If a caller is checking whether a flag that has variants is enabled, the feature manager will check if the variant assigned to the current user is set up to override the result. This is done using the optional variant property `StatusOverride`. By default, this property is set to `None`, which means the variant doesn't affect whether the flag is considered enabled or disabled. Setting `StatusOverride` to `Enabled` allows the variant, when chosen, to override a flag to be enabled. Setting `StatusOverride` to `Disabled` provides the opposite functionality, therefore disabling the flag when the variant is chosen. A feature with a `Status` of `Disabled` cannot be overridden.

Reviewer 2: Maybe we should add a "why?" here, because there's only a narrow reason they would want to use the overrides. It's if a customer wants to allocate via variants without changing their existing IsEnabled code. Meaning they want to return bool results but have more than 2 "buckets".

Author: > Today, feature filters can be used to conditionally enable feature flags. Variants should be able to extend this functionality by also playing a role in the decision to enable a feature flag. A property can be included in a feature flag’s variants giving them the opportunity to override the flag’s enabled state. This is important considering the enabled state of a flag is used in our integration points with ASP.NET Core such as the FeatureGateAttribute

quoted from the [design doc](https://microsoft.sharepoint.com/:w:/r/teams/azconfig/_layouts/15/Doc.aspx?sourcedoc=%7BBA41C52A-B66F-4B9A-A3CA-D5CF6283BE06%7D&file=Variants%20as%20an%20extension%20of%20feature%20flags.docx&action=default&mobileredirect=true&share=IQEqxUG6b7aaS6PK1c9ig74GAQjMj0RepCJ3cjLLW_AeLf0)

This is the original purpose for this design.

> This gives variants an opportunity to extend the evaluation of a feature flag.

I think this sentence is a solid reason but we may come up a good story or a attracting use case of statusOverride.

Reviewer 2: > This is important considering the enabled state of a flag is used in our integration points with ASP.NET Core such as the FeatureGateAttribute

This part is valid, but it's still not a big use case. Essentially, customers can always use a normal flag with filters to achieve whatever boolean targeting they want to achieve. Variants don't offer an advantage to boolean targeting over the standard TargetingFilter. 

One reason someone would use Variants in the bool world is to adding more telemetry "buckets" so they can split their traffic (although this is also not necessary as the split traffic of "true" results should be equal). 

Another reason is if a customer's plan is to call both GetVariant and IsEnabled(like using FeatureGateAttribute) on a single flag, then they might use it, but this is likely overloading what a flag should be in charge of. 

My fear is that users will see this section in the readme and think that they **should** be using it- instead of using it to address a specific problem.

In the above example, the feature is enabled by the `AlwaysOn` filter. If the current user is in the calculated percentile range of 10 to 20, then the `On` variant is returned. Otherwise, the `Off` variant is returned and because `StatusOverride` is equal to `Disabled`, the feature will now be considered disabled.

> ### Variant Service

Reviewer: I would highlight Dependency Injection here.

Something like

## Use variant services in dependency injection

Reviewer: Also, probably makes sense to be in the ToC at the top of the readme.

> Dependency injection can be wired up with the variant feature flag. Different implementations of a service interface can be considered as variant services. Specific variant of an interface can be retrieved from the dependency injection container based on the allocated variant of a variant feature flag. This could be done by using `IVariantServiceProvider<TService>`.

Reviewer 2: ```suggestion
Dependency injection can be wired up with a variant feature flag. Different implementations of a service interface can be considered as variant services. Specific variant of an interface can be retrieved from the dependency injection container based on the allocated variant of a variant feature flag. This could be done by using `IVariantServiceProvider<TService>`.

```

``` C#
    IVariantServiceProvider<IAlgorithm> algorithmServiceProvider;
    ...

    IAlgorithm forecastAlgorithm = await algorithmServiceProvider.GetServiceAsync(cancellationToken); 
```

The `IVaraintServiceProvider<IAlgorithm>` will retrieve a specific implementation of `IAlgorithm` from the dependency injection container. It can be registered by calling `WithvariantService<TService>` on the `IFeatureManagementBuilder`.

``` C#
services.AddFeatureManagement() 
        .WithVariantService<IAlgorithm>("ForecastAlgorithm");
```

With the call above, `IAlgorithm` will be wired up with the variant feature flag `ForecastAlgorithm`. The `GetServiceAsync` method of `IVariantService<IAlgorithm>` will retrieve the variant service of `IAlgorithm` which matches the name of allocated variant of the `ForecastAlgorithm` flag, from the dependency injection container.

``` javascript
{
    "ForecastAlgorithm": {
        "Variants": [ 
            { 
                "Name": "AlgorithmBeta" 
            }
        ] 
    }
}
```

> **Note:** The `WithvariantService<TService>` method will not register any implementation of `TService`. You need to register variant serices as `TService` to allow `IVariantServiceProvider<TService>` to retrieve them from the dependency injection container.

Reviewer 2: ```suggestion
**Note:** The `WithVariantService<TService>` method will not register any implementation of `TService`. You need to register variant serices as `TService` to allow `IVariantServiceProvider<TService>` to retrieve them from the dependency injection container.

```

Reviewer 2: I feel like we should include a snippet of registering the AlgorithmBeta class as defined in this note. Just

```csharp
services.AddSingleton<IAlgorithm, AlgorithmBeta>();
```

Author: User can register AlgorithemBeta in many ways:
``` C#
services.AddSingleton<IAlgorithm, AlgorithmBeta>()
services.AddSingleton<IAlgorithm>(sp => new AlgorithmBeta())
services.AddSingleton<IAlgorithm>(new AlgorithmBeta())
```
I don't want to leave an impression here that AlgorithmBeta has to be registered like this. My thought is that when the example application is ready, we can put a link here and say "For more details, please refer to this example."

Author: How about this:
Which extension method to use for registration depends on your needs. Here is an example.
``` C#
services.AddSingleton<IAlgorithm, AlgorithmBeta>();
```

In the above example, the feature is enabled by the `AlwaysOn` filter. If the current user is in the calculated percentile range of 10 to 20, then the `On` variant is returned. Otherwise, the `Off` variant is returned and because `StatusOverride` is equal to `Disabled`, the feature will now be considered disabled.

> ## Variants in Dependency Injection

Reviewer: Seems like it should be a triple `#` no?

> Dependency injection can be wired up with a variant feature flag. You can use a variant feature flag to control which implementation of a service should be used by the dependency injection, based on the allocated variant of the feature flag. This could be done by using `IVariantServiceProvider<TService>`.

Reviewer: ```suggestion
Variant feature flags can be used in conjunction with dependency injection to surface different implementations of a service for different users. This is accomplished through the use of the `IVariantServiceProvider<TService>` interface.

```

``` C#
IVariantServiceProvider<IAlgorithm> algorithmServiceProvider;
...

IAlgorithm forecastAlgorithm = await algorithmServiceProvider.GetServiceAsync(cancellationToken); 
```

> The `IVaraintServiceProvider<IAlgorithm>` will retrieve a specific implementation of `IAlgorithm` from the dependency injection container. It can be registered by calling `WithVariantService<TService>` on the `IFeatureManagementBuilder`.

Reviewer: ```suggestion
In the snippet above, the `IVariantServiceProvider<IAlgorithm>` will retrieve an implementation of `IAlgorithm` from the dependency injection container. The chosen implementation is dependent upon:
* The feature that the `IAlgorithm` service was registered with
* The allocated variant for that feature

The `IVariantServiceProvider<T>` is made available to the application by calling `IFeatureManagementBuilder.WithVariantService<T>(string featureName)`. See below for an example.

```

``` C#
services.AddFeatureManagement() 
        .WithVariantService<IAlgorithm>("ForecastAlgorithm");
```

With the call above, `IAlgorithm` will be wired up with the variant feature flag `ForecastAlgorithm`. The `GetServiceAsync` method of `IVariantService<IAlgorithm>` will retrieve the variant service of `IAlgorithm` which matches the name of allocated variant of the `ForecastAlgorithm` flag, from the dependency injection container.

``` javascript
{
    "ForecastAlgorithm": {
        "Variants": [
            { 
                "Name": "AlgorithmBeta" 
            },
            ...
        ] 
    }
}
```

> **Note:** The `WithvariantService<TService>` method will not register any implementation of `TService`. You need to register variant serices as `TService` to allow `IVariantServiceProvider<TService>` to retrieve them from the dependency injection container. Which extension method to use for registration depends on your needs. Here is an example.

Reviewer: ```suggestion
**Note:** The `WithVariantService<TService>` method will not register any implementation of `TService`. You need to register variant services as `TService` to allow `IVariantServiceProvider<TService>` to retrieve them from the dependency injection container. Which extension method to use for registration depends on your needs. Here is an example.

```

``` C#

> services.AddSingleton<IAlgorithm, AlgorithmBeta>();

Reviewer: ```suggestion
// Adds an implementation for `IAlgorithm`, which is consumed as a variant service above.
services.AddSingleton<IAlgorithm, AlgorithmBeta>();

```

Reviewer: What happens if we forget to register implementation for the service?

Author: It will return null if there is no matched implementation found in the DI container.

Reviewer: How about

The call above makes `IVariantServiceProvider<IAlgorithm>` available in the service collection. Implementation(s) of `IAlgorithm` must be added separately via an add method such as `services.AddSingleton<IAlgorithm, SomeImplementation>()`. The implementation of `IAlgorithm` that the `IVariantServiceProvider` uses depends on the `ForecastAlgorithm` feature. If no implementation of `IAlgorithm` is added to the service collection, then the `IVariantServiceProvider<IAlgorithm>.GetServiceAsync()` will return a task with a null result.
